uniprotkb-rest/src/test/java/org/uniprot/api/uniprotkb/queue/UniProtKBMessageListenerTest.java

```diff
@@ -0,0 +1,153 @@
+
+        this.uniProtKBMessageListener.onMessage(message);
+
+        // verify the ids file and clean up
+        Path idsFilePath = Path.of("target/" + jobId);
+        Assertions.assertTrue(Files.exists(idsFilePath));
+        List<String> ids = Files.readAllLines(idsFilePath);
+        Assertions.assertNotNull(ids);
+        Assertions.assertEquals(accessions.size(), ids.size());
+        Assertions.assertEquals(accessions, ids);
+        Files.delete(idsFilePath);
+        Assertions.assertTrue(Files.notExists(idsFilePath));
+    }
+
+    @Test
+    void testOnMessageWithIOExceptionDuringWrite() throws IOException {
+        UniProtKBDownloadRequest downloadRequest = new UniProtKBDownloadRequest();
+        downloadRequest.setQuery("field2:value2");
+        MediaType format = MediaType.APPLICATION_JSON;
+        downloadRequest.setFormat(format.toString());
+        String jobId = UUID.randomUUID().toString();
+        MessageBuilder builder = MessageBuilder.withBody(downloadRequest.toString().getBytes());
+        Message message = builder.setHeader("jobId", jobId).build();
+        DownloadJob downloadJob = DownloadJob.builder().id(jobId).build();
+        // stub
+        List<String> accessions = List.of("P12345", "P05067");
+        when(this.jobRepository.findById(jobId)).thenReturn(Optional.of(downloadJob));
+        when(this.converter.fromMessage(message)).thenReturn(downloadRequest);
+        when(this.downloadConfigProperties.getIdFilesFolder()).thenReturn("target");
+        when(this.service.streamIds(downloadRequest)).thenReturn(accessions.stream());
+        Mockito.doThrow(new IOException("Forced IO Exception"))
+                .when(this.downloadResultWriter)
+                .writeResult(any(), any(), any(), any(), any());
+        this.uniProtKBMessageListener.setMaxRetryCount(3);
+
+        this.uniProtKBMessageListener.onMessage(message);
+
+        // verify the ids file is cleaned up during IOException
+        Path idsFilePath = Path.of("target/" + jobId);
+        Assertions.assertTrue(Files.notExists(idsFilePath));
+        Path resultFilePath = Path.of("target/" + jobId + ".json");
+        Assertions.assertTrue(Files.notExists(resultFilePath));
+    }
+
+    @Test
+    void testOnMessageWhenMaxRetryReached() throws IOException {
+        UniProtKBDownloadRequest downloadRequest = new UniProtKBDownloadRequest();
+        downloadRequest.setQuery("field1:value1");
+        downloadRequest.setFormat(MediaType.APPLICATION_JSON.toString());
+        String jobId = UUID.randomUUID().toString();
+        MessageBuilder builder = MessageBuilder.withBody(downloadRequest.toString().getBytes());
+        Message message = builder.setHeader("jobId", jobId).build();
+        DownloadJob downloadJob = DownloadJob.builder().id(jobId).build();
+        this.uniProtKBMessageListener.setMaxRetryCount(0);
+        Assertions.assertDoesNotThrow(() -> this.uniProtKBMessageListener.onMessage(message));
```

Comment:
How do you know that it has returned the maxRetry use case? is there a way to assert it?

Reply:
We are setting the maxretry to 0(it will be 3(> 0) in prod) and the call will enter into if block of  if (isMaxRetriedReached(message)) {}; in the very first call itself.
There is nothing to assert here except that it doen's throw any exception.